Documentation/network/kubernetes/ciliumendpointslice.rst

```diff
@@ -6,28 +6,112 @@
+
+* Make sure that CEPs are enabled (the ``--disable-endpoint-crd`` flag is not set to ``true``)
+* Make sure you are not relying on the Egress Gateway which is not compatible with CES (see Egress Gateway :ref:`egress-gateway-incompatible-features`)
+
+Migration Procedure
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In order to minimize endpoint propagation delays, it is recommended to upgrade the Operator first,
+let it create all CES objects, and then upgrade the Agents afterwards.
+
+#. Enable CES on the Operator by setting the ``enableCiliumEndpointSlice`` value to ``true`` in your Helm chart or
+   by directly setting the ``--enable-cilium-endpoint-slice`` flag to ``true`` on the Operator. Re-deploy the Operator.
+
+#. Once the Operator is running, verify that the ``CiliumEndpointSlice`` CRD has been successfully registered:
+
+   .. code-block:: shell-session
+
+      $ kubectl get crd ciliumendpointslices.cilium.io
+      NAME                                         CREATED AT
+      ciliumendpointslices.cilium.io               2021-11-05T05:41:28Z
+
+#. Verify that the Operator has started creating CES objects:
+
+   .. code-block:: shell-session
+
+      $ kubectl get ces
+      NAME                  AGE
+      ces-2fvynpvzn-4ncg9   1m17s
+      ces-2jyqj8pfl-tdfm8   1m20s
+
+#. Let the Operator create CES objects for all existing CEPs in the cluster. This may take some time, depending on the
+   size of the cluster. You can monitor the progress by checking the rate of CES object creation in the cluster, for example by
+   looking at the ``apiserver_storage_objects`` Kubernetes metric or by looking at ``ciliumendpointslices`` resource
+   creation requests in Kubernetes Audit Logs. You can also monitor the metrics emitted by the Operator, such as ``cilium_operator_ces_sync_total``. All CES-related metrics are documented in the :ref:`ces_metrics` section of the metric documentation.
+
+#. Once the metrics have stabilized (in other words, when the Operator has created CES objects for all existing CEPs), upgrade the
+   Cilium Agents on all nodes by setting the ``--enable-cilium-endpoint-slice`` flag to ``true`` and re-deploying them.
+
+
+Configuration Options
+=====================
+
+Several options are available to adjust the performance and behavior of the CES feature:
+
+* You can configure the way CEPs are batched into CES by changing the maximum number of CEPs in a
+  CES (``--ces-max-cilium-endpoints-per-ces``) or by changing the way CEPs are grouped into CES (``--ces-slice-mode``).
+  Right now two modes are supported: ``cesSliceModeIdentity`` which groups CEPs based on :ref:`security_identities`
+  and ``cesSliceModeFCFS`` which groups CEPs on a "First Come, First Served" basis.
+
+* You can also fine-tune rate-limiting settings for the Operator communications with the API-server. Refer to the ``--ces-*`` flags for the ``cilium-operator`` binary.
```

Comment:
We'll probably want to deprecate the rate limiter flags (especially the manually set ones), because we want to rely on K8s API Priority & Fairness.

Also, there is a k8s client rate limiter that is set for the entire cilium-operator. This affects CES as well.
We first decoupled it from directly limiting CES qps https://github.com/cilium/cilium/pull/24675
We also made a separate k8s client for leader election.

Reply:
Makes sense, I'd say this could be a follow-up item